File: backend/app/models/people.py
# backend/app/models/people.py


"""
PEOPLE TABLE EXAMPLE
--------------------

id                                   | institution_id                       | external_reference | first_name | last_name | preferred_name | email               | lifecycle_stage | attributes                         | created_at                       | updated_at
-------------------------------------|--------------------------------------|--------------------|------------|-----------|----------------|---------------------|-----------------|------------------------------------|------------------------------------|--------------------------------
10000000-0000-4000-8000-000000000001 | aaaaaaaa-aaaa-4aaa-8aaa-aaaaaaaaaaaa | UI-APP-1001        | John       | Smith     | Johnny         | john@example.com    | applicant       | {"source":"website"}               | 2026-08-09 18:50:00.123456-07    | 2026-08-09 18:50:00.123456-07
10000000-0000-4000-8000-000000000002 | aaaaaaaa-aaaa-4aaa-8aaa-aaaaaaaaaaaa | UI-STU-1002        | Mary       | Johnson   | null           | mary@example.com    | enrolled        | {"resident":true,"cohort":"2025"}  | 2026-08-09 18:51:00.123456-07    | 2026-08-09 18:55:00.123456-07

COLUMN MEANING
--------------
id: Unique person UUID.
institution_id: Institution this person belongs to.
external_reference: Identifier from another system (e.g., UI-APP-1001, UI-STU-1002).
first_name: Legal/given first name.
last_name: Family/last name.
preferred_name: Name the person prefers to use.
email: Email address.
lifecycle_stage: Where person is in the CRM lifecycle. Values: prospect, applicant, admitted, committed, enrolled, alumni, inactive.
attributes: Flexible JSON data for additional information.
created_at: Record creation timestamp.
updated_at: Last update timestamp.
UNIQUE RULES: institution_id + email must be unique.
"""

from datetime import date, datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, EmailStr, Field

LifecycleStage = Literal[
    "prospect",
    "applicant",
    "admitted",
    "committed",
    "enrolled",
    "alumni",
    "inactive",
]


class PeopleBase(BaseModel):
    institution_id: UUID
    external_reference: str | None = None
    first_name: str
    last_name: str
    preferred_name: str | None = None
    email: EmailStr | None = None
    phone: str | None = None
    date_of_birth: date | None = None
    lifecycle_stage: LifecycleStage = "prospect"
    attributes: dict[str, Any] = Field(default_factory=dict)


class PeopleCreate(PeopleBase):
    pass


class PeopleUpdate(BaseModel):
    external_reference: str | None = None
    first_name: str | None = None
    last_name: str | None = None
    preferred_name: str | None = None
    email: EmailStr | None = None
    phone: str | None = None
    date_of_birth: date | None = None
    lifecycle_stage: LifecycleStage | None = None
    attributes: dict[str, Any] | None = None


class PeopleResponse(PeopleBase):
    id: UUID
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/models/people.js
// src/models/people.js

class People {
  constructor(
    id,
    institution_id,
    external_reference,
    first_name,
    last_name,
    preferred_name,
    email,
    phone,
    date_of_birth,
    lifecycle_stage,
    attributes,
    created_at,
    updated_at
  ) {
    this.id = id;
    this.institution_id = institution_id;
    this.external_reference = external_reference;
    this.first_name = first_name;
    this.last_name = last_name;
    this.preferred_name = preferred_name;
    this.email = email;
    this.phone = phone;
    this.date_of_birth = date_of_birth;
    this.lifecycle_stage = lifecycle_stage;
    this.attributes = attributes;
    this.created_at = created_at;
    this.updated_at = updated_at;
  }

  // Helper getters
  get fullName() {
    return `${this.first_name} ${this.last_name}`;
  }

  get displayName() {
    return this.preferred_name || this.fullName;
  }

  get institutionId() {
    return this.institution_id;
  }

  static fromApiResponse(data) {
    if (!data) return null;
    return new People(
      data.id,
      data.institution_id ?? data.institutionId,
      data.external_reference,
      data.first_name,
      data.last_name,
      data.preferred_name,
      data.email,
      data.phone,
      data.date_of_birth ?? data.dateOfBirth,
      data.lifecycle_stage ?? data.lifecycleStage,
      data.attributes,
      data.created_at ?? data.createdAt,
      data.updated_at ?? data.updatedAt
    );
  }

  toJSON() {
    return {
      id: this.id,
      institution_id: this.institution_id,
      external_reference: this.external_reference,
      first_name: this.first_name,
      last_name: this.last_name,
      preferred_name: this.preferred_name,
      email: this.email,
      phone: this.phone,
      date_of_birth: this.date_of_birth,
      lifecycle_stage: this.lifecycle_stage,
      attributes: this.attributes,
      created_at: this.created_at,
      updated_at: this.updated_at,
    };
  }

  createPayload() {
    return {
      institution_id: this.institution_id,
      external_reference: this.external_reference,
      first_name: this.first_name,
      last_name: this.last_name,
      preferred_name: this.preferred_name,
      email: this.email,
      phone: this.phone,
      date_of_birth: this.date_of_birth,
      lifecycle_stage: this.lifecycle_stage,
      attributes: this.attributes,
    };
  }
}

export default People;
